framework/data_types/ndarray.h

```diff
@@ -446,94 +242,104 @@ class NDArray
   template <typename D>
   void resize(const std::initializer_list<D>& dims)
   {
-    static_assert(std::is_integral<D>::value,
-                  "NDArray dims argument must have vector of integral types.");
-
-    dimensions_.clear();
-    strides_.clear();
-    delete[] base_;
-    base_ = nullptr;
-
-    NDArray<T>(dims).swap(*this);
+    set_dimensions(dims);
+    if (size_ != std::accumulate(dims.begin(), dims.end(), 1, std::multiplies<size_t>()))
+      initialize();
   }
 
   /**
-   * Accesses the specified element.
+   * Accesses the specified element for an array with N dimensions.
    *
-   * \param args The indices of the desired element.
+   * \param args Indices for each dimension.
    * \return Read/write reference to the element.
    */
   template <typename... Args>
-  T& operator()(Args... args) noexcept
+  inline __attribute__((always_inline)) constexpr T& operator()(Args... args) noexcept
   {
-    static_assert(AllIntegral<Args...>::value,
-                  "NDArray::operator[]: All parameters must be of integral type");
+    return storage_[compute_index(args...)];
+  }
 
+  /**
+   * Accesses the specified element for an array with N dimensions with bounds checking.
+   *
+   * \param args Indices for each dimension.
+   * \return Read/write reference to the element.
+   */
+  template <typename... Args>
+  constexpr T& at(Args... args)
+  {
     size_t indices[]{static_cast<size_t>(args)...};
-
-    const size_t N = rank_;
-
-    T* address = base_ + indices[N - 1];
-    for (size_t i = 0; i < N - 1; ++i)
-      address += strides_[i] * indices[i];
-    return *(address);
+    for (size_t i = 0; i < rank_; ++i)
+    {
+      if (indices[i] >= dimensions_[i])
+        throw std::out_of_range("Index out of bounds.");
+    }
+    return (*this)(args...);
```

Comment:
```suggestion
    return storage[compute_index(args...)];
```
Just a suggestion that you can ignore. Never been a big fan of calling `()` in this way within a class.

Reply:
Fixed.